feat(plugin): User Voice Show

LordEliasTM opened the pull request:
Shows in which voice channel a user is.
If you click the button with the channel name you can join the voice if you have sufficient permissions to the channel.
Strongly inspired by the BetterDiscord version of this plugin.

![image](https://user-images.githubusercontent.com/55048695/227657237-aa88109c-e393-4068-9156-10ec961a1c38.png)
![image](https://user-images.githubusercontent.com/55048695/227657128-e01961b5-d054-45d6-9d00-2fdf5f4d5d04.png)

Vendicated commented:
The ui for this looks very foreign, it doesn't fit into Discord at all. I would make it look similar to activities. I think that's what the BD plugin does? Perhaps you could look at it for inspiration

LordEliasTM commented:
I just copy pasted the style from the original plugin. However what do you mean with the activities?

AutumnVN commented:
i think he means this https://betterdiscord.app/plugin/VoiceActivity

![](https://raw.githubusercontent.com/Neodymium7/BetterDiscordStuff/main/VoiceActivity/assets/preview.png)

LordEliasTM commented:
Well to make it more fitting into discord, I'd make it a normal button instead of custom. However it'd look weird on the user profile modal.

LordEliasTM commented:
Also I intended the design to be more subtile, as the custom button matches the background when not hovered over.

LordEliasTM commented:
> I think that's what the BD plugin does?

> i think he means this https://betterdiscord.app/plugin/VoiceActivity

I based it on https://betterdiscord.app/plugin/UserVoiceShow, not VoiceActivity

LordEliasTM commented:
Hope that looks native enough.
Added a title and changed my custom button to a native one.
![image](https://user-images.githubusercontent.com/55048695/227808547-7535c8de-94ac-4b18-b626-db12d09e8a37.png)

Vendicated commented on src/plugins/userVoiceShow/components/VoiceChannelSection.tsx:
```diff
@@ -0,0 +1,64 @@
+            style={{
+                width: "100%",
+                margin: "auto",
+                height: "unset",
```
why's this here, couldn't this be in the css file instead?

Vendicated commented on src/plugins/userVoiceShow/components/VoiceChannelSection.tsx:
```diff
@@ -0,0 +1,64 @@
+
+const CONNECT = 1n << 20n;
+
+type VoiceChannelFieldProps = {
```
```suggestion
interface VoiceChannelFieldProps {
```

Vendicated commented on src/plugins/userVoiceShow/components/VoiceChannelSection.css:
```diff
@@ -0,0 +1,33 @@
+    font-family: var(--font-display);
+    font-weight: 700;
+    text-transform: uppercase;
+}
```
please use em instead of px. Also instead of manually styling this header, you can just use Forms.FormTitle instead

Vendicated commented on src/plugins/userVoiceShow/components/VoiceChannelSection.css:
```diff
@@ -0,0 +1,33 @@
+    text-align: center;
+    padding: 5px;
+    color: #fff !important;
+    font-size: 16px !important;
```
avoid using important so users can more easily style it

Vendicated commented on src/plugins/userVoiceShow/index.tsx:
```diff
@@ -0,0 +1,113 @@
+    },
+    showVoiceChannelSectionHeader: {
+        type: OptionType.BOOLEAN,
+        description: "Wether to show \"IN A VOICE CHANNEL\" above the join button",
```
```suggestion
        description: 'Whether to show "IN A VOICE CHANNEL" above the join button',
```

Vendicated commented on src/plugins/userVoiceShow/index.tsx:
```diff
@@ -0,0 +1,113 @@
+
+export default definePlugin({
+    name: "User Voice Show",
+    description: "See the channel a user is sitting in and click join that channel",
```
```suggestion
    description: "Shows whether a User is currently in a voice channel somewhere in their profile",
```

LordEliasTM replied:
How do i make it obvious that you can join the channel by clicking then?

Vendicated commented on src/plugins/userVoiceShow/index.tsx:
```diff
@@ -0,0 +1,113 @@
+    }
+});
+
+const getVoiceChannelField = (props: any) => {
```
React Components should use an upper case name. void is not a valid return type for React components, return null instead!

Vendicated commented on src/plugins/userVoiceShow/index.tsx:
```diff
@@ -0,0 +1,113 @@
+    patchModal(props: any) {
+        if (settings.store.showInUserProfileModal)
+            return (
+                <div style={{ margin: "0 12px" }} >
```
this will unconditionally add a random margin, even if the user is not connected to any voice channel

LordEliasTM replied:
i think it wouldn't even matter, because the height is 0, changed it anyway

Vendicated commented on src/plugins/userVoiceShow/index.tsx:
```diff
@@ -0,0 +1,113 @@
+                // $1: argument name (currently "e")
+                // $2: the rest inbetween the argument name and my actual match
+                // $3: my actual match
+                match: /(?<=function \w+\()(\w)(.*)(\(0,\w\.jsx\))(?=(.(?!\3))+?canDM)/,
```
both your regular expressions are extremely unsafe and probably also poorly performing. Try avoiding dangerous matchers like .* and this use of backreferences is likely not needed either. If you show me the code in question and how you want to modify it, I can help you write a better matcher

LordEliasTM replied:
i found some safer regexes, but i hardcoded "e" as the argument name of that original function. Can you think of a way to also capture the argument name in case it changes?

Nuckyz commented on src/plugins/userVoiceShow/index.tsx:
```diff
@@ -0,0 +1,107 @@
+            replacement: {
+                match: /,{user:\w{1,2}}\)(?!;case)/,
+                // paste my fancy custom button below the username
+                replace: "$&,$self.patchModal(e)",
```
```suggestion
                replace: "$&,$self.patchModal(arguments[0])",
```

pinetreemoonlight commented:
Hey, hope it's fine that I'm posting here. I have a feature request about your awesome plugin:

1. incorporate friend list voice activity icon
2. incorporate individual user's DM panel right-hand side panel voice activity

These two features are present in BetterDiscord's VoiceActivity plugin by Neodymium: https://github.com/Neodymium7/BetterDiscordStuff/blob/main/VoiceActivity/VoiceActivity.plugin.js

Would this be possible?

Here are pics:
Friend list voice activity:
![Screenshot 2024-04-30 at 06 42 28](https://github.com/Vendicated/Vencord/assets/115622756/c4bf204e-97ce-4c57-90e1-e66c2ef88a20)

DM right-hand side panel voice activity:
![Screenshot 2024-04-30 at 06 42 56](https://github.com/Vendicated/Vencord/assets/115622756/d5766dd2-246f-4f5e-bcf3-ee4fd6603554)

D3SOX commented:
@pinetreemoonlight This was already asked a thousand times in Discord and there are existing plugin-requests

https://github.com/Vencord/plugin-requests/issues/173
https://github.com/Vencord/plugin-requests/issues/141

and PRs
- #1669
- #1783
- #2248

for this

LordEliasTM commented:
Exactly, I've already continued plugin development in private, as the Vencord devs seem to ignore my pull requests, so I just gave up. Already added features like a button to just view the channel and a speaker icon next to names of friends who are in a voice.

D3SOX commented:
I gave up waiting for my PRs as well, but I provide them as userplugins now. It would be great if you could do that with the latest version of your plugin too

pinetreemoonlight commented:
> Sure, [userVoiceShow2.zip](https://github.com/Vendicated/Vencord/files/15164234/userVoiceShow2.zip)

![image](https://github.com/Vendicated/Vencord/assets/115622756/fdb7e9b4-da49-42af-bfa2-9b4c06ca48a0)

I tried to use it, but I can't see voice icons in the member list even though it's enabled in the settings.

LordEliasTM commented:
Bruh, any errors?

pinetreemoonlight commented:
> Bruh, any errors?

ok so it works great on member/friends list, but not on individual DM's right-hand side panel

practiceineverything commented:
> Конечно, [userVoiceShow2.zip](https://github.com/Vendicated/Vencord/files/15164234/userVoiceShow2.zip)

Can you please tell me how I can install it?

practiceineverything commented:
> Bruh, any errors?

It's not working at the moment, can you fix it please.)
